src/mlir/generator.cc

```diff
@@ -31,29 +31,31 @@ namespace mlir::verona
   using namespace ASTInterface;
 
   // ===================================================== Public Interface
-  void Generator::readAST(const ::ast::Ast& ast)
+  llvm::Error Generator::readAST(const ::ast::Ast& ast)
   {
     // Parse the AST with the rules below
-    parseModule(ast);
+    if (auto err = parseModule(ast))
+      return err;
 
     // On error, dump module for debug purposes
     if (mlir::failed(mlir::verify(*module)))
     {
       module->dump();
-      throw std::runtime_error("MLIR verification failed from Verona file");
+      return runtimeError("MLIR verification failed from Verona file");
     }
+    return llvm::Error::success();
   }
 
-  void Generator::readMLIR(const std::string& filename)
+  llvm::Error Generator::readMLIR(const std::string& filename)
   {
     if (filename.empty())
-      throw std::runtime_error("No input filename provided");
+      return runtimeError("No input filename provided");
 
     // Read an MLIR file
     auto srcOrErr = llvm::MemoryBuffer::getFileOrSTDIN(filename);
 
     if (auto err = srcOrErr.getError())
-      throw std::runtime_error(
+      return runtimeError(
         "Cannot open file " + filename + ": " + err.message());
```

Comment:
We should not be doing string concatenation like this if this is a human-facing error message.  The existing error reporting code uses the `fmt` library to collect format strings and their arguments so that they can be localised and rendered for humans.